graphjet-core/src/main/java/com/twitter/graphjet/algorithms/socialproof/EntitySocialProofGenerator.java

```diff
@@ -0,0 +1,162 @@
+
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectArrayMap;
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectMap;
+import it.unimi.dsi.fastutil.bytes.ByteArraySet;
+import it.unimi.dsi.fastutil.bytes.ByteSet;
+import it.unimi.dsi.fastutil.ints.Int2DoubleMap;
+import it.unimi.dsi.fastutil.ints.Int2DoubleOpenHashMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.ints.IntSet;
+import it.unimi.dsi.fastutil.longs.Long2DoubleMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectArrayMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
+import it.unimi.dsi.fastutil.longs.LongArraySet;
+import it.unimi.dsi.fastutil.longs.LongSet;
+
+/**
+ * SocialProofGenerator shares similar logic with {@link com.twitter.graphjet.algorithms.counting.TopSecondDegreeByCount}.
+ * In the request, clients specify a seed user set (left nodes) and an entity set (right nodes).
+ * SocialProofGenerator finds the intersection between the seed users' (left node) edges and the given entity set.
+ * Only entities with at least one social proof will be returned to clients.
+ */
+public abstract class EntitySocialProofGenerator implements
+    RecommendationAlgorithm<EntitySocialProofRequest, SocialProofResponse> {
+
+  private static final int MAX_EDGES_PER_NODE = 500;
+  private static final Byte2ObjectMap<Long2ObjectMap<LongSet>> EMPTY_SOCIALPROOF_MAP =
+      new Byte2ObjectArrayMap<>();
+
+  private NodeMetadataLeftIndexedMultiSegmentBipartiteGraph nodeMetadataLeftIndexedBipartiteGraph;
+  // Entity Int -> Engagements Byte -> User Long -> Tweets Seq[Long]
+  private final Int2ObjectMap<Byte2ObjectMap<Long2ObjectMap<LongSet>>> socialProofs;
+  // Tweet Long -> Sum of social proof edges Double
+  private final Int2DoubleMap socialProofWeights;
+  protected RecommendationType recommendationType;
+  protected IDMask idMask;
+
+  public EntitySocialProofGenerator(
+      NodeMetadataLeftIndexedMultiSegmentBipartiteGraph nodeMetadataLeftIndexedBipartiteGraph
+  ) {
+    this.nodeMetadataLeftIndexedBipartiteGraph = nodeMetadataLeftIndexedBipartiteGraph;
+    // Variables socialProofs and socialProofWeights are re-used for each request.
+    this.socialProofs = new Int2ObjectOpenHashMap<>();
+    this.socialProofWeights = new Int2DoubleOpenHashMap();
+  }
+
+  /**
+   * Collect social proofs for a given {@link SocialProofRequest}.
+   *
+   * @param request contains a set of input ids and a set of seed users.
+   */
+  private void collectRecommendations(EntitySocialProofRequest request) {
+    socialProofs.clear();
+    socialProofWeights.clear();
+    IntSet inputEntityIds = request.getEntityIds();
+    ByteSet socialProofTypes = new ByteArraySet(request.getSocialProofTypes());
+
+    // Iterate through the set of seed users with weights. For each seed user, we go through his edges.
+    for (Long2DoubleMap.Entry entry: request.getLeftSeedNodesWithWeight().long2DoubleEntrySet()) {
```

Comment:
nitpicking on style: prefer smaller functions.
- Each loop here can be abstracted into a separate function to make the main loop more readable
- Guard clause can reduce the number of nesting. ex: 
`  if (edgeIterator == null) { 
      continue; 
    }
    <regular logic>
`